File: iosApp/iosApp/ViewModels/ContentViewModel.swift
//
//  ContentViewModel.swift
//  iosApp
//
//  Created by Brady, Kayla on 7/19/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Foundation
@_spi(Experimental) import MapboxMaps
import Shared

class ContentViewModel: ObservableObject {
    @Published var featurePromosPending: [FeaturePromo]?
    @Published var onboardingScreensPending: [OnboardingScreen]?
    @Published var defaultTab: DefaultTab?

    var mapboxConfigManager: IMapboxConfigManager
    var featurePromoUseCase: IFeaturePromoUseCase
    var onboardingRepository: IOnboardingRepository
    var tabPreferencesRepository: ITabPreferencesRepository

    init(mapboxConfigManager: IMapboxConfigManager = MapboxConfigManager(),
         featurePromoUseCase: IFeaturePromoUseCase = UsecaseDI().featurePromoUsecase,
         featurePromosPending: [FeaturePromo]? = nil,
         onboardingRepository: IOnboardingRepository = RepositoryDI().onboarding,
         onboardingScreensPending: [OnboardingScreen]? = nil,
         tabPreferencesRepository: ITabPreferencesRepository = RepositoryDI().tabPreferences) {
        self.mapboxConfigManager = mapboxConfigManager
        self.featurePromoUseCase = featurePromoUseCase
        self.featurePromosPending = featurePromosPending
        self.onboardingRepository = onboardingRepository
        self.onboardingScreensPending = onboardingScreensPending
        self.tabPreferencesRepository = tabPreferencesRepository
    }

    @MainActor func loadConfig() async {
        await mapboxConfigManager.loadConfig()
    }

    @MainActor func loadPendingFeaturePromosAndTabPreferences() async {
        let promos = await (try? featurePromoUseCase.getFeaturePromos()) ?? []
        featurePromosPending = promos
        await loadTabPreferences(promos.contains(where: { $0 == .enhancedFavorites }))
    }

    @MainActor func loadOnboardingScreens() async {
        onboardingScreensPending = await (try? onboardingRepository.getPendingOnboarding()) ?? []
    }

    @MainActor func loadTabPreferences(_ hasPendingFavoritesPromo: Bool) async {
        if hasPendingFavoritesPromo {
            defaultTab = .favorites
            await (try? tabPreferencesRepository.setDefaultTab(defaultTab: .favorites))
        } else {
            defaultTab = await (try? tabPreferencesRepository.getDefaultTab())
        }
    }
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/ContentViewModel.kt
package com.mbta.tid.mbta_app.android

import androidx.lifecycle.ViewModel
import com.mbta.tid.mbta_app.model.FeaturePromo
import com.mbta.tid.mbta_app.model.OnboardingScreen
import com.mbta.tid.mbta_app.repositories.DefaultTab
import com.mbta.tid.mbta_app.repositories.IOnboardingRepository
import com.mbta.tid.mbta_app.repositories.ITabPreferencesRepository
import com.mbta.tid.mbta_app.usecases.IFeaturePromoUseCase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class ContentViewModel(
    private val featurePromoUseCase: IFeaturePromoUseCase,
    private val onboardingRepository: IOnboardingRepository,
    private val tabPreferencesRepository: ITabPreferencesRepository,
) : ViewModel() {
    private val _pendingFeaturePromos: MutableStateFlow<List<FeaturePromo>?> =
        MutableStateFlow(null)
    val pendingFeaturePromos: StateFlow<List<FeaturePromo>?> = _pendingFeaturePromos
    private val _pendingOnboarding: MutableStateFlow<List<OnboardingScreen>?> =
        MutableStateFlow(null)

    val pendingOnboarding: StateFlow<List<OnboardingScreen>?> = _pendingOnboarding

    private val _defaultTab: MutableStateFlow<DefaultTab?> = MutableStateFlow(null)
    val defaultTab: StateFlow<DefaultTab?> = _defaultTab

    init {
        loadPendingFeaturePromosAndTabPreferences()
        loadPendingOnboarding()
    }

    fun loadPendingFeaturePromosAndTabPreferences() {
        CoroutineScope(Dispatchers.IO).launch {
            val data = featurePromoUseCase.getFeaturePromos()
            _pendingFeaturePromos.value = data
            loadTabPreferences(data.contains(FeaturePromo.EnhancedFavorites))
        }
    }

    fun loadPendingOnboarding() {
        CoroutineScope(Dispatchers.IO).launch {
            val data = onboardingRepository.getPendingOnboarding()
            _pendingOnboarding.value = data
        }
    }

    fun loadTabPreferences(hasPendingFavoritesPromo: Boolean) {
        CoroutineScope(Dispatchers.IO).launch {
            if (hasPendingFavoritesPromo) {
                _defaultTab.value = DefaultTab.Favorites
                tabPreferencesRepository.setDefaultTab(DefaultTab.Favorites)
            } else {
                _defaultTab.value = tabPreferencesRepository.getDefaultTab()
            }
        }
    }

    fun clearPendingOnboarding() {
        _pendingOnboarding.value = emptyList()
    }

    fun clearPendingFeaturePromos() {
        _pendingFeaturePromos.value = emptyList()
    }
}
